common/archiver/filestore/historyArchiver.go

```diff
@@ -0,0 +1,284 @@
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+const (
+	fileStoreScheme = "file://"
+
+	archiveNonRetriableErrorMsg = "Archive method encountered an non-retriable error."
+
+	errInvalidURI               = "archive URI is invalid"
+	errInvalidRequest           = "archive request is invalid"
+	errConstructHistoryIterator = "failed to construct history iterator"
+	errReadHistory              = "failed to read history batches"
+	errHistoryMutated           = "history was mutated"
+	errEncodeHistory            = "failed to encode history batches"
+	errMakeDirectory            = "failed to make directory"
+	errWriteFile                = "failed to write history to file"
+)
+
+var (
+	errContextTimeout           = errors.New("activity aborted because context timed out")
+	errInvalidGetHistoryRequest = &shared.BadRequestError{Message: "Get archived history request is invalid"}
+	errGetHistoryTokenCorrupted = &shared.BadRequestError{Message: "Next page token is corrupted."}
+	errHistoryNotExist          = &shared.BadRequestError{Message: "Requested workflow history does not exist."}
+)
+
+type (
+	// HistoryArchiverConfig configs the filestore implementation of archiver.HistoryArchiver interface
+	HistoryArchiverConfig struct {
+		*archiver.HistoryIteratorConfig
+	}
+
+	historyArchiver struct {
+		historyManager   persistence.HistoryManager
+		historyV2Manager persistence.HistoryV2Manager
+		logger           log.Logger
+		config           *HistoryArchiverConfig
+
+		// only set in test code
+		historyIterator archiver.HistoryIterator
+	}
+
+	getHistoryToken struct {
+		CloseFailoverVersion int64
+		NextBatchIdx         int
+	}
+)
+
+func newHistoryArchiver(
+	historyManager persistence.HistoryManager,
+	historyV2Manager persistence.HistoryV2Manager,
+	logger log.Logger,
+	config *HistoryArchiverConfig,
+	historyIterator archiver.HistoryIterator,
+) *historyArchiver {
+	return &historyArchiver{
+		historyManager:   historyManager,
+		historyV2Manager: historyV2Manager,
+		logger:           logger,
+		config:           config,
+		historyIterator:  historyIterator,
+	}
+}
+
+func (h *historyArchiver) Archive(
+	ctx context.Context,
+	URI string,
+	request *archiver.ArchiveHistoryRequest,
+	opts ...archiver.ArchiveOption,
+) error {
+	logger := tagLoggerWithArchiveHistoryRequest(h.logger, request)
+
+	if !h.ValidateURI(URI) {
+		logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errInvalidURI), tag.ArchivalURI(URI))
+		return archiver.ErrArchiveNonRetriable
+	}
+
+	if err := validateArchiveRequest(request); err != nil {
+		logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errInvalidRequest), tag.Error(err))
+		return archiver.ErrArchiveNonRetriable
+	}
+
+	historyIterator := h.historyIterator
+	if historyIterator == nil {
```

Comment:
In addition to the comment on the field above, lets add a comment here indicating that this is due to test injection.